### DISCUSSION TITLE
Automation - Bash scripting documentation

### DISCUSSION THREAD

**Discussion by SantoshSrinivas79** [Budibase/budibase#3359] [Help]
Hello, Is there any documentation on how the bash scripting action in the automation section works?
Please advise anything to get started with it. Thank you!

**Comment by brandonhon** [comment]
@SantoshSrinivas79 What are you trying to do. I can try to help. I have been messing with the bash scripting action in automate for some things I am working on.

Are you running budibase via docker-compose?

**Reply by SantoshSrinivas79** [reply]
Hey @brandonhon, yes ... I'm running Budibase via `docker-compose`. I just need a few samples and use-cases to get started.
Anything you can share would be great, however basic it may be. Just getting started from zero on bash scripting.

**Reply by brandonhon** [reply]
One thing I noticed right away was that the containers are running Alpine Linux. Alpine does not use bash as the default shell instead it uses ash. Ash and Bash are not 100% compatible but many of the basic commands work exactly the same.

To get a shell in the main apps container 'bbapps' use this command. After playing around a little bit you should get a better feel for ash and what it can do. 
`docker exec -it bbapps ash`

From with in the bash scripting action under automate you can run shell commands like `which ls` or `pwd`. This will help you find where things are located and where you should place scripts.

Here is a very simple example you can test with.
App Action: Add a generic field and name it anything. Don't matter what you put here. You can even delete it after adding.
Query Rows: Select your table and setup the other options. For limit put 1. This will return one row from the db table you selected.
Bash Scripting: For code add `echo {{ steps.1.rows }}`.

Run a test and your output should show you the info in the row that was returned from the query rows action.

If you would like to run python or something from the bash scripting action you will need add those packages to the container.  If this is something you are interested in let me know and I can provide you the dockerfile and docker-compose.yml I am using.

**Reply by SantoshSrinivas79** [reply]
Thanks, @brandonhon tried it out. Look interesting. I wish actions could return data that can be used in the App.

Python makes it very interesting .... would be great if you could provide the docker file. Thank you!

**Reply by brandonhon** [reply]
@SantoshSrinivas79 here you go. I moved the app container off of alpine and on to debian. This way it has bash just like the name of the action 'Bash Scripting' 😃.

Dockerfile:
```
FROM node:14-buster-slim

LABEL com.centurylinklabs.watchtower.lifecycle.pre-check="scripts/watchtower-hooks/pre-check.sh"
LABEL com.centurylinklabs.watchtower.lifecycle.pre-update="scripts/watchtower-hooks/pre-update.sh"
LABEL com.centurylinklabs.watchtower.lifecycle.post-update="scripts/watchtower-hooks/post-update.sh"
LABEL com.centurylinklabs.watchtower.lifecycle.post-check="scripts/watchtower-hooks/post-check.sh"

WORKDIR /app

# copy files and install dependencies
ADD . ./
RUN yarn
RUN apt update && apt upgrade -y && apt install -y python3 python3-pip git pdftk curl vim
RUN pip3 install pdfrw pdfforms pyexcel
RUN mkdir /mnt/data

EXPOSE 4001

# have to add node environment production after install
# due to this causing yarn to stop installing dev dependencies
# which are actually needed to get this environment up and running
ENV NODE_ENV=production
CMD ["yarn", "run:docker"]
```


docker-compose.yml change:
```
app-service:
  restart: always
  # image: budibase.docker.scarf.sh/budibase/apps
  build:
    context: .
    dockerfile: Dockerfile
  image: budibase/apps-custom
  container_name: bbapps
```

**Reply by CyberSardinha** [reply]
Can you use a file that is saved as an attachment on the db on the python script?
